### PR TITLE
Add groups to dynamo benchmarking output data

### PR DESCRIPTION
**Pull request by drisspg**
# Summary
Ads the required information to enable this issue:
https://github.com/pytorch/test-infra/issues/4268

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @msaroufim @aakhundov

### CODE REVIEW

File: benchmarks/dynamo/common.py [hunk 1]
```diff
@@ -287,6 +288,31 @@ class CI(NamedTuple):
...
 
+# Model Groups
+@dataclass(frozen=True)
+class ModelGroup:
+    name: str
+    models: Set[str]
+
+
+BLUEBERRY_MODELS = {"llama", "nanogpt_generate"}
+BLUEBERRY_MODELS = ModelGroup("Blueberries", BLUEBERRY_MODELS)
```
Comment (drisspg) [inline benchmarks/dynamo/common.py:299]: fix this driss

**Comment by ezyang** [comment]
This csv format seems incompatible with a model being in multiple groups (which, in the data model you've defined here, is eminently possible). Additionally, what exactly is going to happen to the group data when it hits HUD? In particular, does grouping a model remove it from the default set?

**Comment by drisspg** [comment]
> This csv format seems incompatible with a model being in multiple groups (which, in the data model you've defined here, is eminently possible). Additionally, what exactly is going to happen to the group data when it hits HUD? In particular, does grouping a model remove it from the default set?

@ezyang  you are right, I updated to assert single group membership. TBH I am not sure though if this group thing is going to be extended or used anywhere else. If we need multi-group membership should that just be all "Groups" with a formatted string. I am flying a little blind on the consumer end. 

> Additionally, what exactly is going to happen to the group data when it hits HUD? In particular, does grouping a model remove it from the default set?

@huydhn  Do you have any guidance here? For instance I ran on  "llama", which is part of a group  and "BERT_pytorch" which isn' and  I produce a performance CSV:
Screenshot 2023-06-09 at 9 50 41 AM

Does this provide enough, and the right information to update the hud?

**Comment by huydhn** [comment]
> Does this provide enough, and the right information to update the hud?

I think it does provide enough information to update HUD.  At first, the new group name like BlueBerries won't be used for anything as the compilation, mode, and suite information are extracted from the CSV file name, for example `aot_eager_inference_torchbench`.  So by querying group name in addition to the rest of the information in the CSV, HUD would know which model belong to which group or the default group (empty).

**Comment by huydhn** [comment]
UX-wise, do you envision the `BlueBerries` group to be on par with TorchBench, HuggingFace, and TIMM when displaying on https://hud.pytorch.org/benchmark/compilers? Or is this ok to be part of [TorchBench detail page](https://hud.pytorch.org/benchmark/torchbench/inductor_dynamic?startTime=Sat,%2003%20Jun%202023%2000:01:09%20GMT&stopTime=Sat,%2010%20Jun%202023%2000:01:09%20GMT&granularity=hour&mode=training&dtype=amp&lBranch=main&lCommit=2961ea80f5d5fbcb9ef86d95ce08f1fa603863bf&rBranch=main&rCommit=8215468870daae5a1da7ac02f98f4e2ddf2d7623) for example?

The latter is easier to implement given the current state of this PR as the group name could be easily treated as a property of the model like other CSV columns.

On the other hand, if you prefer the former approach, it would be better to duplicate and store Blueberries models into separate CSV files, for example `inductor_dynamic_blueberries_amp_inference_cuda_performance.csv` and `inductor_dynamic_blueberries_amp_inference_cuda_accuracy.csv`.  Then it would show up on the dashboard as a new suite.  It's ok to also keep the group name as a CSV column.

File: benchmarks/dynamo/common.py [hunk 2]
```diff
@@ -689,8 +726,9 @@ def maybe_mark_profile(*args, **kwargs):
             timings,
         )
 
-    first_headers = ["dev", "name", "batch_size"]
-    first_fields = [current_device, current_name, current_batch_size]
+    current_group = ALL_MODEL_GROUPS.model_group(current_name)
+    first_headers = ["dev", "name", "group", "batch_size"]
```
Comment (huydhn) [inline benchmarks/dynamo/common.py:730]: In addition to the new column, is this possible to filter the result dataframe and store Blueberies models in separate CSV files such as `inductor_dynamic_blueberries_amp_inference_cuda_performance.csv` and `inductor_dynamic_blueberries_amp_inference_cuda_accuracy.csv` with `blueberries` in the file name.  Duplicated records are fine, so llama, for example, could appear in both TorchBench CSV file and Blueberries CSV file

**Comment by ezyang** [comment]
So, here's my take about optimal UX.

The *simplest* UX change is to treat blueberries as a split configuration on torchbench: blueberries is another benchmark suite on par with torchbench/huggingface/timm, and we're splitting out the models from torchbench into blueberries. Because it is annoying to have torchbench.py and blueberries.py, you would augment the benchmark script to output multiple csvs, as huydn suggested.

The *logical* UX change, in my opinion, is that benchmark groups are an overlay *on top* of the torchbench/huggingface/timm benchmark suites. In other words, it's a way to collect up results from torchbench/huggingface/timm and aggregate the in a non-standard way. In this universe, I would NOT make any changes to the benchmark suite; instead, we would code the list of blueberries models in the HUD, and the HUD would be entirely responsible for also doing the side calculations to give these extra aggregates. If we modify the blueberries group, we can retroactively put it more results for models that hadn't previously been included.

**Comment by ezyang** [comment]
If it wasn't clear, I prefer the latter, but that's more work on @huydhn

**Comment by huydhn** [comment]
> The _logical_ UX change, in my opinion, is that benchmark groups are an overlay _on top_ of the torchbench/huggingface/timm benchmark suites. In other words, it's a way to collect up results from torchbench/huggingface/timm and aggregate the in a non-standard way. In this universe, I would NOT make any changes to the benchmark suite; instead, we would code the list of blueberries models in the HUD, and the HUD would be entirely responsible for also doing the side calculations to give these extra aggregates. If we modify the blueberries group, we can retroactively put it more results for models that hadn't previously been included.

Yup, I'm in line with this.  Although putting the list of blueberries models on HUD requires some further thoughts as we probably don't want it to be hard coded.  So the mapping info would still need to be on Rockset (doesn't need to be on the same perf benchmark table though)

IMO, the former (simple) and the latter (logical) changes don't seem like mutually exclusive options.  A few more records on Rockset under blueberries sounds fine to me in case they are needed.  So I think the way to move forward quickly would be to just output multiple CSV files first (if that's easily done), so there are records on Rockset.  The UX change could follow right after.

**Comment by huydhn** [comment]
If outputting multiple CSV files makes the script look terrible and brittle (not sure how it will look), please let me know.  I would be happy to stamp the PR as it stands because having the group name already provides the required mapping information.

**Review by huydhn — APPROVED** [review]

**Pull request closed without merging**